src/main/java/mate/academy/spring/service/impl/CustomUserDetailsServiceImpl.java

```diff
@@ -0,0 +1,36 @@
+package mate.academy.spring.service.impl;
+
+import static org.springframework.security.core.userdetails.User.withUsername;
+
+import java.util.Optional;
+import mate.academy.spring.model.User;
+import mate.academy.spring.service.UserService;
+import org.springframework.security.core.userdetails.User.UserBuilder;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
+@Service
+public class CustomUserDetailsServiceImpl implements UserDetailsService {
+    private final UserService userService;
+
+    public CustomUserDetailsServiceImpl(UserService userService) {
+        this.userService = userService;
+    }
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        Optional<User> userOptional = userService.findByEmail(email);
+        if (userOptional.isEmpty()) {
+            throw new UsernameNotFoundException("User not found for this email: " + email);
+        }
+        User user = userOptional.get();
+        UserBuilder builder = withUsername(email);
+        builder.password(user.getPassword());
+        builder.roles(user.getRoles().stream()
+                .map(role -> role.getRoleName().name())
+                .toArray(String[]::new));
+        return builder.build();
```

Comment:
Usually we use builders in such way:
```suggestion
                String[] roles = user.getRoles().stream()
                .map(Role::getRoleName)
                .toArray(String[]::new));
        return .withUsername(email)
                .password(user.getPassword())
                .roles(roles)
                .build();
```